datalad/cmd.py

```diff
@@ -6,66 +6,86 @@
     """
 
     __slots__ = ['commands', 'dry']
 
     def __init__(self, dry=False):
+        self.dry = dry
         self.commands = []
-        self.dry = dry                    # TODO: make use of it
 
-    # TODO -- remove dry_run specification here -- use constructor parameter
-    #def __call__(self, cmd, dry_run=False):
-    def getstatusoutput(self, cmd, dry_run=None):
-        """A wrapper around commands.getstatusoutput
+    def run(self, cmd):
+        """Runs the command `cmd` using shell.
+
+        `cmd` is called and uses stdout whereas stderr is captured and logged.
+        In case of dry-mode `cmd` is just added to `commands`.
 
-        Provides improved logging for debugging purposes and raises
-        RuntimeError exception in case of non-0 return
+        Parameters
+        ----------
+        cmd : str
+          String defining the command call.
+
+        Returns
+        -------
+        Status code as returned by the called command or `None` in case of dry-mode.
         """
-        self.log("Running: %s" % (cmd,))
+
+        self.log("Running: %s" % cmd)
+
         if not self.dry:
-            # status, output = commands.getstatusoutput(cmd)
-            # doing manually for improved debugging
-            pipe = os.popen('{ ' + cmd + '; } 2>&1', 'r')
-            output = ''
-            for line in iter(pipe.readline, ''):
-                self.log("| " + line.rstrip('\n'), level=5)
-                output += line
-            if output[-1:] == '\n': output = output[:-1]
-            status = pipe.close() or 0
+
+            proc = subprocess.Popen(cmd, stdout=sys.stdout, stderr=subprocess.PIPE, shell=True)
+            # shell=True allows for piping, multiple commands, etc., but that implies to not use shlex.split()
+            # and is considered to be a security hazard. So be careful with input.
+            # Alternatively we would have to parse `cmd` and create multiple subprocesses.
+
+            while proc.poll() is None:
+                err_line = proc.stderr.readline()
+                if err_line != '':
+                    self.log("stderr: %s" % err_line)
+                    #TODO: log per line or as a whole?
+
+            status = proc.poll()
 
             if not status in [0, None]:
-                msg = "Failed to run %r. Exit code=%d output=%s" \
-                      % (cmd, status, output)
+                msg = "Failed to run %r. Exit code=%d" % (cmd, status)
                 lgr.error(msg)
                 raise RuntimeError(msg)
+
             else:
-                self.log("Finished running %r with status %s" % (cmd, status),
-                         level=8)
-                return status, output
+                self.log("Finished running %r with status %s" % (cmd, status), level=8)
+                return status
+
         else:
             self.commands.append(cmd)
-        return None, None
+        return None
 
     def drycall(self, f, *args, **kwargs):
```

Comment:
as you expressed dislike of the name (Which I don't like either):  name it ```call()```?
Then also it could provide ```__call__``` which would call run or call depending on the argument (str vs a callable) ?